Rubycritic compare branches

HemanthMudalaiah opened the pull request.

Commit:
This commit will have changes in rubycritic where you can compare two branches

Commit:
This commit has changes where we can decide which of the file changes degraded the code quality

Commit:
Comparing the diff with the whole branch for duplication analysis

Commit:
Configuring smell link to point to feature branch code

Commit:
Changes in the content of the build details

Commit:
Gitlab Integration

Pushing the status of the code and report location to gitlab using gitlab api

Commit:
Code cleanup

* Code refactoring, DRY
* Comparing between branches in local without pr id.

Commit:
Specs and code cleanup

1. Setting a new mode to compare between branches.
2. Moved the code changes to a seperate file.
3. Pushing comments on to pr.
4. Specs for analysed_module_collection.

tejasbubane commented on lib/rubycritic/cli/options.rb:
```diff
@@ -71,14 +79,18 @@ def to_h
 
       attr_accessor :mode, :root, :format, :deduplicate_symlinks,
-                    :suppress_ratings, :minimum_score, :no_browser, :parser
+                    :suppress_ratings, :minimum_score, :no_browser, :parser, :base_branch, :feature_branch, :compare_between_branches, :merge_request_id
```
line too long, break it

tejasbubane commented on lib/rubycritic/cli/options.rb:
```diff
@@ -19,6 +19,14 @@ def parse
+            self.feature_branch = String(branches.split(',')[1].strip)
+            self.merge_request_id = Integer(branches.split(',')[2].strip) rescue nil
+            self.compare_between_branches = true
+            self.mode = :compare_branches
```
The two options `compare_between_branches` and `mode = :compare_branches` are redundant. Keep `mode` only and check for `mode == :compare_branches` or better yet add methods like `Config.compare_branches_mode?` and similar for other modes.

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,203 @@
+        @files_affected = []
+        @analysed_modules
+        @number
+        @code_index_file_location
```
why these initializations without assignment?

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,203 @@
+        report(critic)
+        Config.feature_branch_flag = false
+        switch_branch(Config.base_branch)
+      end
```
Code duplication in above two methods `switch_to_base_branch_and_compare` and `switch_to_feature_branch_and_compare`

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,203 @@
+            (Config.base_branch_score - Config.feature_branch_score).round(2)
+          else
+            (Config.feature_branch_score - Config.base_branch_score).round(2)
+          end
```
Remove `if` and use `abs` instead.

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,203 @@
+
+      def compare_code_quality
+        Config.base_branch_score > Config.feature_branch_score ? Config.quality_flag = false : Config.quality_flag = true
+        status_reporter.status_message = Config.quality_flag ? "GOOOOOOOOD" : "BAAAAAAAD #{@files_affected} files degraded."
```
Was this `Good` / `Bad` present in previous code? If not remove, we don't want to change stdout unless absolutely required.

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,203 @@
+      end
+
+      def mark_jenkins_build_fail
+        [@app_settings['app_threshold'], @app_settings['difference_threshold']].all?(&:nil?) ? false : (Config.base_branch_score < @app_settings['app_threshold'] || (Config.base_branch_score - Config.feature_branch_score) > @app_settings['difference_threshold']) 
```
Line too long and complex.

tejasbubane replied:
Also check this for thresholds, does your code override this or is it separate? https://github.com/cybrilla/rubycritic/blob/master/docs/building-own-code-climate.md#failing-fast-the---minimum-score-option

tejasbubane commented on lib/rubycritic/configuration.rb:
```diff
@@ -3,7 +3,8 @@ module RubyCritic
     attr_accessor :source_control_system, :mode, :format, :deduplicate_symlinks,
-                  :suppress_ratings, :open_with, :no_browser
+                  :suppress_ratings, :open_with, :no_browser, :base_branch, :feature_branch, :compare_between_branches ,:base_branch_score, :feature_branch_score, :quality_flag,
+                  :base_root_directory, :feature_root_directory, :build_root_directory, :set_location, :base_branch_flag, :feature_branch_flag, :build_flag, :merge_request_id, :difference_score
```
break lines

tejasbubane commented on lib/rubycritic/configuration.rb:
```diff
@@ -13,6 +14,14 @@ def set(options)
+      self.set_location = false
+      self.feature_branch_flag = false
+      self.base_branch_flag = false
+      self.build_flag = false
```
Do we need these flags?

tejasbubane commented on lib/rubycritic/core/analysed_modules_collection.rb:
```diff
@@ -16,16 +16,27 @@ class AnalysedModulesCollection
+      unless modules.nil?
+        @modules = SourceLocator.new(paths).pathnames.map do |pathname|
+          analysed_module = modules.find { |mod|  mod.pathname == pathname }
+          AnalysedModule.new(pathname: analysed_module.pathname, name: analysed_module.name, smells: analysed_module.smells, churn: analysed_module.churn, committed_at: analysed_module.committed_at, complexity: analysed_module.complexity, duplication: analysed_module.duplication, methods_count: analysed_module.methods_count)
```
line break - try to follow 80 characters limit for line length.

tejasbubane commented on lib/rubycritic/generators/html/templates/layouts/application.html.erb:
```diff
@@ -11,11 +11,18 @@
           <a href="<%= file_path('smells_index.html') %>" class="project-nav-item">Smells</a>
+          <% if Config.compare_between_branches %>
+            &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp
+            &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp &nbsp 
```
Use css margin/padding instead of lots of `&nbsp`s

tejasbubane commented on test/lib/rubycritic/commands/compare_test.rb:
```diff
@@ -0,0 +1,35 @@
+      @options[:merge_request_id].must_equal 1
+    end
+  end
+end
```
I think more tests are required for this mode. This just checks if the `config` is being set. But what happens after that? Does the `compare` mode run correctly analysing both branches and create htmls accordingly?

tejasbubane commented:
We should remove the `gitlab/github` commenting part from this and just keep this to running rubycritic on two branches. That commenting really does not belong to `rubycritic` (that's why I was thinking of making a separate gem for that). But since the `CI mode` makes sense in rubycritic and the maintainer had added it to the roadmap, we are pushing this in.

tejasbubane commented:
@vijayanandnandam Can you enable `CodeClimate` and `Coveralls` for this cloned repo so that we make sure the code is good before submitting upstream?

Commit:
Code Cleanup & Refactoring

1. Removed Gitlab,Github integrations.
2. Template Changes.
3. Options changes.

Commit:
Added specs and threshold option

1. Added specs for compare module
2. Added threshold as an option in rubycritic before it was as app side configuration
3. Check for any uncommited changes

Commit:
Removing simplecov from dependency

Commit:
Refactoring compare module

1. Removed redundant code.
2. Adding config so that browser doesn't open html report during test.

tejasbubane commented:
@HemanthMudalaiah Why did you close this?
